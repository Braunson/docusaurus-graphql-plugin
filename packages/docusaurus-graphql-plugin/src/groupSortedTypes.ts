import {
  GraphQLNamedType,
  isEnumType,
  isInputObjectType,
  isInterfaceType,
  isObjectType,
  isScalarType,
  isUnionType,
} from "graphql";
import { GroupedTypes, TypeItem } from "./types";

function groupTypes(types: GraphQLNamedType[]): GroupedTypes {
  const groups = types.reduce<GroupedTypes>(
    (acc, type) => {
      if (type.name.startsWith("__")) {
        return acc;
      }

      if (isObjectType(type)) {
        if (type.name === "Query") {
          return {
            ...acc,
            queries: Object.values(type.getFields()).map((query) => ({
              type: query,
            })),
          };
        }

        if (type.name === "Mutation") {
          return {
            ...acc,
            mutations: Object.values(type.getFields()).map((mutation) => ({
              type: mutation,
            })),
          };
        }

        return {
          ...acc,
          objects: acc.objects.concat([{ type }]),
        };
      }

      if (isInterfaceType(type)) {
        return {
          ...acc,
          interfaces: acc.interfaces.concat([
            {
              type,
              implementedBy: [],
            },
          ]),
        };
      }

      if (isEnumType(type)) {
        return {
          ...acc,
          enums: acc.enums.concat([{ type }]),
        };
      }

      if (isUnionType(type)) {
        return {
          ...acc,
          unions: acc.unions.concat([{ type }]),
        };
      }

      if (isInputObjectType(type)) {
        return {
          ...acc,
          inputObjects: acc.inputObjects.concat([{ type }]),
        };
      }

      if (isScalarType(type)) {
        return {
          ...acc,
          scalars: acc.scalars.concat([{ type }]),
        };
      }

      return acc;
    },
    {
      queries: [],
      mutations: [],
      objects: [],
      interfaces: [],
      enums: [],
      unions: [],
      inputObjects: [],
      scalars: [],
    }
  );

  [...groups.objects, ...groups.interfaces].forEach(({ type }) => {
    type.getInterfaces().forEach((inter) => {
      groups.interfaces.forEach((otherInter) => {
        if (otherInter.type.name === inter.name) {
          otherInter.implementedBy.push(type);
        }
      });
    });
  });

  return groups;
}

function sortByName<T extends TypeItem<{ name: string }>>(types: T[]): T[] {
  return types.sort((a, b) => a.type.name.localeCompare(b.type.name));
}

function sortGroupedTypes(groupedTypes: GroupedTypes): GroupedTypes {
  return {
    queries: sortByName(groupedTypes.queries),
    mutations: sortByName(groupedTypes.mutations),
    objects: sortByName(groupedTypes.objects),
    interfaces: sortByName(groupedTypes.interfaces),
    enums: sortByName(groupedTypes.enums),
    unions: sortByName(groupedTypes.unions),
    inputObjects: sortByName(groupedTypes.inputObjects),
    scalars: sortByName(groupedTypes.scalars),
  };
}

export function groupSortedTypes(types: GraphQLNamedType[]): GroupedTypes {
  return sortGroupedTypes(groupTypes(types));
}
