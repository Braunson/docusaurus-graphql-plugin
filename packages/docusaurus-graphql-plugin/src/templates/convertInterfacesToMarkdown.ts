import { InterfaceItem, MarkdownOptions } from "../types";
import { pushFields } from "./pushFields";
import { pushInterfaces } from "./pushInterfaces";

export function convertInterfacesToMarkdown(
  interfaces: InterfaceItem[],
  { getTypePath }: MarkdownOptions
): string {
  const lines: string[] = [];

  interfaces.forEach(({ type: inter, implementedBy }) => {
    lines.push(`## ${inter.name}`, `\n\n`);
    lines.push(inter.description || "", `\n\n`);

    const subInterfaces = inter.getInterfaces();
    if (subInterfaces.length > 0) {
      pushInterfaces(lines, subInterfaces, { getTypePath });
    }

    if (implementedBy.length > 0) {
      lines.push(
        `<p style={{ marginBottom: "0.4em" }}><strong>Implemented by</strong></p>`,
        `\n\n`
      );
      implementedBy.forEach((object) => {
        lines.push(`- [${object.name}](${getTypePath(object)})`, `\n`);
      });
      lines.push(`\n`);
    }

    const fields = Object.values(inter.getFields());
    pushFields(lines, fields, { getTypePath });
  });

  return lines.join("");
}
