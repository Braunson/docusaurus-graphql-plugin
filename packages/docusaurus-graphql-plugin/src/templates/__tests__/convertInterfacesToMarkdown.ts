import { buildSchema } from "graphql";
import { groupSortedTypes } from "../../groupSortedTypes";
import { convertInterfacesToMarkdown } from "../convertInterfacesToMarkdown";

describe("convertInterfacesToMarkdown", () => {
  function getTypePath() {
    return "#";
  }

  it("should render interfaces", () => {
    const schema = buildSchema(`
      """
      Properties that are common to the different kind of users.
      """
      interface User {
        "The user's name."
        name: String!
      }
    `);
    const { interfaces } = groupSortedTypes(Object.values(schema.getTypeMap()));

    expect(
      convertInterfacesToMarkdown(interfaces, {
        getTypePath,
      })
    ).toMatchSnapshot();
  });

  it("should render the interface's interfaces", () => {
    const schema = buildSchema(`
      """
      Properties that are common to the different kind of users.
      """
      interface User {
        "The user's name."
        name: String!
      }

      """
      A special kind of user.
      """
      interface Admin implements User {
        "The admin's special power."
        power: String!
      }
    `);
    const { interfaces } = groupSortedTypes(Object.values(schema.getTypeMap()));

    expect(
      convertInterfacesToMarkdown(interfaces, {
        getTypePath,
      })
    ).toMatchSnapshot();
  });

  it("should list the interface's implementations", () => {
    const schema = buildSchema(`
      """
      Properties that are common to the different kind of users.
      """
      interface User {
        "The user's name."
        name: String!
      }
      
      """
      A special kind of user.
      """
      type Admin implements User {
        "The admin's name."
        name: String!
      }
    `);
    const { interfaces } = groupSortedTypes(Object.values(schema.getTypeMap()));

    expect(
      convertInterfacesToMarkdown(interfaces, {
        getTypePath,
      })
    ).toMatchSnapshot();
  });
});
