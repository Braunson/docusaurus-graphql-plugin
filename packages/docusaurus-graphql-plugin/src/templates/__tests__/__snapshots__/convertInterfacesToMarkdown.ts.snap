// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertInterfacesToMarkdown should list the interface's implementations 1`] = `
"## User

Properties that are common to the different kind of users.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Implemented by</strong></p>

- [Admin](#)

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Fields</strong></p>

<table>
<thead><tr><th>Name</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>
name<br />
<a href=\\"#\\"><code>String!</code></a>
</td>
<td>
<p>The user&#39;s name.</p>
</td>
</tr>
</tbody>
</table>

"
`;

exports[`convertInterfacesToMarkdown should render interfaces 1`] = `
"## User

Properties that are common to the different kind of users.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Fields</strong></p>

<table>
<thead><tr><th>Name</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>
name<br />
<a href=\\"#\\"><code>String!</code></a>
</td>
<td>
<p>The user&#39;s name.</p>
</td>
</tr>
</tbody>
</table>

"
`;

exports[`convertInterfacesToMarkdown should render the interface's interfaces 1`] = `
"## Admin

A special kind of user.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Implements</strong></p>

- [User](#)

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Fields</strong></p>

<table>
<thead><tr><th>Name</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>
power<br />
<a href=\\"#\\"><code>String!</code></a>
</td>
<td>
<p>The admin&#39;s special power.</p>
</td>
</tr>
</tbody>
</table>

## User

Properties that are common to the different kind of users.

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Implemented by</strong></p>

- [Admin](#)

<p style={{ marginBottom: \\"0.4em\\" }}><strong>Fields</strong></p>

<table>
<thead><tr><th>Name</th><th>Description</th></tr></thead>
<tbody>
<tr>
<td>
name<br />
<a href=\\"#\\"><code>String!</code></a>
</td>
<td>
<p>The user&#39;s name.</p>
</td>
</tr>
</tbody>
</table>

"
`;
